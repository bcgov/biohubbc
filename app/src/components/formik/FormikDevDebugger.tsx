import { Button } from '@mui/material';
import { useFormikContext } from 'formik';
import { useState } from 'react';

/**
 * Renders formik form values. Used for manual testing purposes
 * Only renders display button when in development.
 * Can provide a custom payload / object for additional testing.
 *
 * @params {FormikDevDebugger}
 * @returns {*}
 *
 **/

interface FormikDevDebuggerProps {
  custom_payload?: any;
}
const FormikDevDebugger = ({ custom_payload }: FormikDevDebuggerProps) => {
  const { values, errors, touched } = useFormikContext();
  const [showFormDebugger, setShowDebugger] = useState(false);
  return (
    <>
      {/* Only render the button in Development */}
      {process.env.NODE_ENV === 'development' ? (
        <Button onClick={() => setShowDebugger((d) => !d)}>Formik Debugger (Only Dev)</Button>
      ) : null}
      {showFormDebugger ? (
        <pre>
          {JSON.stringify(
            { FORMIK_ERRORS: errors, FORMIK_VALUES: values, FORMIK_TOUCHED: touched, CUSTOM_PAYLOAD: custom_payload },
            null,
            2
          )}
        </pre>
      ) : null}
    </>
  );
};

export default FormikDevDebugger;
