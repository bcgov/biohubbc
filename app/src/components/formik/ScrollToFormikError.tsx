import Snackbar from '@material-ui/core/Snackbar';
import Alert from '@material-ui/lab/Alert';
import { useFormikContext } from 'formik';
import { IGetProjectForViewResponse } from 'interfaces/useProjectApi.interface';
import React, { useEffect, useState } from 'react';

export interface IScrollToFormikErrorProps {
  /**
   * An ordered list of field names, which informs which field this component will scroll to when multiple fields are
   * in error.
   *
   * Note: A regex is required if the field name has dynamic components (ie: a field with an array of objects where
   * part of the field name is its index in the array field).
   *
   * @type {((string | RegExp)[])}
   * @memberof IScrollToFormikErrorProps
   */
  fieldOrder: (string | RegExp)[];
}

export const ScrollToFormikError: React.FC<IScrollToFormikErrorProps> = (props) => {
  const formikProps = useFormikContext<IGetProjectForViewResponse>();
  const { errors, submitCount } = formikProps;
  const [openSnackbar, setOpenSnackbar] = useState({ open: false, msg: '' });

  useEffect(() => {
    const showSnackBar = (message: string) => {
      setOpenSnackbar({ open: true, msg: message });
    };

    const getAllFieldErrorNames = (obj: object, prefix = '', result: string[] = []) => {
      Object.keys(obj).forEach((key) => {
        const value = obj[key];
        if (!value) return;

        key = Number(key) || key === '0' ? `[${key}]` : key;

        const nextKey = prefix ? `${prefix}.${key}` : key;

        if (typeof value === 'object') {
          getAllFieldErrorNames(value, nextKey, result);
        } else {
          result.push(nextKey);
        }
      });
      return result;
    };

    const getFirstErrorField = (errorArray: string[]): string | undefined => {
      for (const listError of props.fieldOrder) {
        for (const trueError of errorArray) {
          if (trueError.match(listError) || listError === trueError) {
            return trueError;
          }
        }
      }
    };

    const fieldErrorNames = getAllFieldErrorNames(errors);

    const topFieldError = getFirstErrorField(fieldErrorNames);

    if (!topFieldError) {
      return;
    }

    showSnackBar(`Missing one or more required fields.`);

    const errorElement = document.getElementsByName(topFieldError);

    if (errorElement.length <= 0) {
      return;
    }

    errorElement[0].scrollIntoView({ behavior: 'smooth', block: 'center' });

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [errors, submitCount]);

  const closeSnackBar = () =>
    setOpenSnackbar(() => {
      return { open: false, msg: '' };
    });

  return (
    <>
      <Snackbar
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center'
        }}
        open={openSnackbar.open}
        onClose={closeSnackBar}>
        <Alert elevation={4} severity="error" variant="filled" onClose={closeSnackBar}>
          {openSnackbar.msg}
        </Alert>
      </Snackbar>
    </>
  );
};
