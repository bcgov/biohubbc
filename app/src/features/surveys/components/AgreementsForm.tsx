import Box from '@material-ui/core/Box';
import Checkbox from '@material-ui/core/Checkbox';
import FormControl from '@material-ui/core/FormControl';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import FormHelperText from '@material-ui/core/FormHelperText';
import Grid from '@material-ui/core/Grid';
import makeStyles from '@material-ui/core/styles/makeStyles';
import Typography from '@material-ui/core/Typography';
import { useFormikContext } from 'formik';
import React from 'react';
import yup from 'utils/YupSchema';

const useStyles = makeStyles({
  bold: {
    fontWeight: 'bold'
  }
});

export interface IAgreementsForm {
  agreements: {
    sedis_procedures_accepted: boolean;
    foippa_requirements_accepted: boolean;
  };
}

export const AgreementsInitialValues: IAgreementsForm = {
  agreements: {
    sedis_procedures_accepted: false,
    foippa_requirements_accepted: false
  }
};

export const AgreementsYupSchema = yup.object().shape({
  agreements: yup.object().shape({
    sedis_procedures_accepted: yup.bool().oneOf([true], 'You must agree to the SEDIS procedures'),
    foippa_requirements_accepted: yup.bool().oneOf([true], 'You must agree to the FOIPPA requirements')
  })
});

/**
 * Create survey - agreements fields
 *
 * @return {*}
 */
const AgreementsForm = () => {
  const classes = useStyles();

  const { errors, touched, values, handleChange } = useFormikContext<IAgreementsForm>();

  return (
    <form>
      <Grid container spacing={3}>
        <Grid item xs={12}>
          <Typography className={classes.bold}>
            Species and Ecosystems Data and Information Security (SEDIS) Procedures
          </Typography>
        </Grid>
        <Grid item xs={12}>
          <FormControl
            required={true}
            component="fieldset"
            error={
              touched.agreements?.sedis_procedures_accepted && Boolean(errors.agreements?.sedis_procedures_accepted)
            }>
            <FormControlLabel
              control={
                <Checkbox
                  checked={values.agreements?.sedis_procedures_accepted}
                  onChange={handleChange}
                  name="agreements.sedis_procedures_accepted"
                  color="primary"
                />
              }
              label="All data and information for this survey will be collected legally, and in accordance with Section 1 of the Species and Ecosystems Data and Information Security (SEDIS) procedures."
            />
            <FormHelperText>{errors.agreements?.sedis_procedures_accepted}</FormHelperText>
          </FormControl>
        </Grid>
        <Grid item xs={12}>
          <Box pt={4}>
            <Typography className={classes.bold}>
              Freedom of Information and Protection of Privacy Act (FOIPPA) Requirements
            </Typography>
          </Box>
        </Grid>
        <Grid item xs={12}>
          <FormControl
            required={true}
            component="fieldset"
            error={
              touched.agreements?.foippa_requirements_accepted &&
              Boolean(errors.agreements?.foippa_requirements_accepted)
            }>
            <FormControlLabel
              control={
                <Checkbox
                  checked={values.agreements?.foippa_requirements_accepted}
                  onChange={handleChange}
                  name="agreements.foippa_requirements_accepted"
                  color="primary"
                />
              }
              label="All data submitted in this survey will meet or exceed the Freedom of Information and Protection of Privacy Act (FOIPPA) requirements."
            />
            <FormHelperText>{errors.agreements?.foippa_requirements_accepted}</FormHelperText>
          </FormControl>
        </Grid>
      </Grid>
    </form>
  );
};

export default AgreementsForm;
