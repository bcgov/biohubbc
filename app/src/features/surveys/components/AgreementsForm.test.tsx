import { render } from '@testing-library/react';
import { Formik } from 'formik';
import AgreementsForm, {
  AgreementsInitialValues,
  AgreementsYupSchema
} from 'features/surveys/components/AgreementsForm';
import React from 'react';

const handleSaveAndNext = jest.fn();

const agreementsFilledValues = {
  sedis_procedures_accepted: true,
  foippa_requirements_accepted: true
};

describe('Agreements Form', () => {
  it('renders correctly the empty component correctly', () => {
    const { asFragment } = render(
      <Formik
        initialValues={AgreementsInitialValues}
        validationSchema={AgreementsYupSchema}
        validateOnBlur={true}
        validateOnChange={false}
        onSubmit={async (values) => {
          handleSaveAndNext(values);
        }}>
        {() => <AgreementsForm />}
      </Formik>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly the filled component correctly', () => {
    const { asFragment } = render(
      <Formik
        initialValues={agreementsFilledValues}
        validationSchema={AgreementsYupSchema}
        validateOnBlur={true}
        validateOnChange={false}
        onSubmit={async (values) => {
          handleSaveAndNext(values);
        }}>
        {() => <AgreementsForm />}
      </Formik>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
