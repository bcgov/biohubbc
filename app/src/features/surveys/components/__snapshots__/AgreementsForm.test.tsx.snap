// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Agreements Form renders correctly the empty component correctly 1`] = `
<DocumentFragment>
  <form>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <p
          class="MuiTypography-root makeStyles-bold-1 MuiTypography-body1"
        >
          Species and Ecosystems Data and Information Security (SEDIS) Procedures
        </p>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <fieldset
          class="MuiFormControl-root"
        >
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-2 MuiCheckbox-root MuiCheckbox-colorPrimary MuiIconButton-colorPrimary"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  class="PrivateSwitchBase-input-5"
                  data-indeterminate="false"
                  name="sedis_procedures_accepted"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              All data and information for this survey will be collected legally, and in accordance with Section 1 of the Species and Ecosystems Data and Information Security (SEDIS) procedures.
            </span>
          </label>
          <p
            class="MuiFormHelperText-root Mui-required"
          />
        </fieldset>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <div
          class="MuiBox-root MuiBox-root-6"
        >
          <p
            class="MuiTypography-root makeStyles-bold-1 MuiTypography-body1"
          >
            Freedom of Information and Protection of Privacy Act (FOIPPA) Requirements
          </p>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <fieldset
          class="MuiFormControl-root"
        >
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-2 MuiCheckbox-root MuiCheckbox-colorPrimary MuiIconButton-colorPrimary"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  class="PrivateSwitchBase-input-5"
                  data-indeterminate="false"
                  name="foippa_requirements_accepted"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              All data submitted in this survey will meet or exceed the Freedom of Information and Protection of Privacy Act (FOIPPA) requirements.
            </span>
          </label>
          <p
            class="MuiFormHelperText-root Mui-required"
          />
        </fieldset>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`Agreements Form renders correctly the filled component correctly 1`] = `
<DocumentFragment>
  <form>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <p
          class="MuiTypography-root makeStyles-bold-7 MuiTypography-body1"
        >
          Species and Ecosystems Data and Information Security (SEDIS) Procedures
        </p>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <fieldset
          class="MuiFormControl-root"
        >
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-8 MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-checked-9 Mui-checked MuiIconButton-colorPrimary"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  checked=""
                  class="PrivateSwitchBase-input-11"
                  data-indeterminate="false"
                  name="sedis_procedures_accepted"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              All data and information for this survey will be collected legally, and in accordance with Section 1 of the Species and Ecosystems Data and Information Security (SEDIS) procedures.
            </span>
          </label>
          <p
            class="MuiFormHelperText-root Mui-required"
          />
        </fieldset>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <div
          class="MuiBox-root MuiBox-root-12"
        >
          <p
            class="MuiTypography-root makeStyles-bold-7 MuiTypography-body1"
          >
            Freedom of Information and Protection of Privacy Act (FOIPPA) Requirements
          </p>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <fieldset
          class="MuiFormControl-root"
        >
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-8 MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-checked-9 Mui-checked MuiIconButton-colorPrimary"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  checked=""
                  class="PrivateSwitchBase-input-11"
                  data-indeterminate="false"
                  name="foippa_requirements_accepted"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              All data submitted in this survey will meet or exceed the Freedom of Information and Protection of Privacy Act (FOIPPA) requirements.
            </span>
          </label>
          <p
            class="MuiFormHelperText-root Mui-required"
          />
        </fieldset>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
